File: content/realm/examples/ios/Examples/Relationships.m
// :replace-start: {
//   "terms": {
//     "InverseRelationshipObjcExample_": "",
//     "ToManyObjcExample_": "",
//     "ToOneRelationshipObjc_": ""
//   }
// }
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>

// :snippet-start: inverse-relationship
// InverseRelationshipObjcExample_Task.h
@interface InverseRelationshipObjcExample_Task : RLMObject
@property NSString *description;
@property (readonly) RLMLinkingObjects *assignees;
@end

// Define an RLMArray<InverseRelationshipObjcExample_Task> type
RLM_COLLECTION_TYPE(InverseRelationshipObjcExample_Task)


// InverseRelationshipObjcExample_User.h
@interface InverseRelationshipObjcExample_User : RLMObject
@property NSString *name;
@property RLMArray<InverseRelationshipObjcExample_Task *><InverseRelationshipObjcExample_Task> *tasks;
@end


// InverseRelationshipObjcExample_Task.m
@implementation InverseRelationshipObjcExample_Task
+ (NSDictionary *)linkingObjectsProperties {
    return @{
        @"assignees": [RLMPropertyDescriptor descriptorWithClass:InverseRelationshipObjcExample_User.class propertyName:@"tasks"],
    };
}
@end


// InverseRelationshipObjcExample_User.m
@implementation InverseRelationshipObjcExample_User
@end

// :snippet-end:

// :snippet-start: to-many-relationship
// ToManyObjcExample_Dog.h
@interface ToManyObjcExample_Dog : RLMObject
@property NSString *name;
// No backlink to person -- one-directional relationship
@end

// Define an RLMArray<ToManyObjcExample_Dog> type
RLM_COLLECTION_TYPE(ToManyObjcExample_Dog)

// ToManyObjcExample_Person.h
@interface ToManyObjcExample_Person : RLMObject
@property NSString *name;
// A person can have many dogs
@property RLMArray<ToManyObjcExample_Dog *><ToManyObjcExample_Dog> *dogs;
@end


// ToManyObjcExample_Dog.m
@implementation ToManyObjcExample_Dog
@end

// ToManyObjcExample_Person.m
@implementation ToManyObjcExample_Person
@end
// :snippet-end:

// :snippet-start: to-one-relationship
// ToOneRelationshipObjc_Dog.h
@interface ToOneRelationshipObjc_Dog : RLMObject
@property NSString *name;
// No backlink to person -- one-directional relationship
@end

// Define an RLMArray<ToManyObjcExample_Dog> type
RLM_COLLECTION_TYPE(ToOneRelationshipObjc_Dog)

// ToOneRelationshipObjc_Person.h
@interface ToOneRelationshipObjc_Person : RLMObject
@property NSString *name;
// A person can have one dog
@property ToOneRelationshipObjc_Dog *dog;
@end


// ToOneRelationshipObjc_Dog.m
@implementation ToOneRelationshipObjc_Dog
@end

// ToOneRelationshipObjc_Person.m
@implementation ToOneRelationshipObjc_Person
@end
// :snippet-end:

@interface RelationshipsObjc : XCTestCase
@end

@implementation RelationshipsObjc
@end

// :replace-end:

File: content/realm/examples/ios/Examples/Relationships.swift
// :replace-start: {
//   "terms": {
//     "InverseRelationshipExample_": "",
//     "InverseRelationshipObjcDynamicExample_": "",
//     "ToManyExample_": "",
//     "ToOneRelationship_": ""
//   }
// }
import Foundation
import RealmSwift

// :snippet-start: inverse-relationship
class InverseRelationshipExample_User: Object {
    @Persisted(primaryKey: true) var _id: ObjectId
    @Persisted var _partition: String = ""
    @Persisted var name: String = ""

    // A user can have many tasks.
    @Persisted var tasks: List<InverseRelationshipExample_Task>
}

class InverseRelationshipExample_Task: Object {
    @Persisted(primaryKey: true) var _id: ObjectId
    @Persisted var _partition: String = ""
    @Persisted var text: String = ""

    // Backlink to the user. This is automatically updated whenever
    // this task is added to or removed from a user's task list.
    @Persisted(originProperty: "tasks") var assignee: LinkingObjects<InverseRelationshipExample_User>
}
// :snippet-end:

// :snippet-start: inverse-relationship-objc-dynamic
class InverseRelationshipObjcDynamicExample_User: Object {
    @objc dynamic var _id: ObjectId = ObjectId.generate()
    @objc dynamic var _partition: String = ""
    @objc dynamic var name: String = ""

    // A user can have many tasks.
    let tasks = List<InverseRelationshipObjcDynamicExample_Task>()

    override static func primaryKey() -> String? {
        return "_id"
    }
}

class InverseRelationshipObjcDynamicExample_Task: Object {
    @objc dynamic var _id: ObjectId = ObjectId.generate()
    @objc dynamic var _partition: String = ""
    @objc dynamic var text: String = ""

    // Backlink to the user. This is automatically updated whenever
    // this task is added to or removed from a user's task list.
    let assignee = LinkingObjects(fromType: InverseRelationshipObjcDynamicExample_User.self, property: "tasks")

    override static func primaryKey() -> String? {
        return "_id"
    }
}
// :snippet-end:

// :snippet-start: to-many-relationship
class ToManyExample_Person: Object {
    @Persisted var name: String = ""
    @Persisted var birthdate: Date = Date(timeIntervalSince1970: 1)

    // A person can have many dogs
    @Persisted var dogs: List<ToManyExample_Dog>
}

class ToManyExample_Dog: Object {
    @Persisted var name: String = ""
    @Persisted var age: Int = 0
    @Persisted var breed: String?
    // No backlink to person -- one-directional relationship
}
// :snippet-end:

// :snippet-start: to-one-relationship
class ToOneRelationship_Person: Object {
    @Persisted var name: String = ""
    @Persisted var birthdate: Date = Date(timeIntervalSince1970: 1)

    // A person can have one dog
    @Persisted var dog: ToOneRelationship_Dog?
}

class ToOneRelationship_Dog: Object {
    @Persisted var name: String = ""
    @Persisted var age: Int = 0
    @Persisted var breed: String?
    // No backlink to person -- one-directional relationship
}
// :snippet-end:

// :replace-end:
